File: src/main/java/org/revcloud/vader/lift/InheritanceLiftUtil.java
package org.revcloud.vader.lift;

import io.vavr.control.Either;
import lombok.experimental.UtilityClass;
import org.revcloud.vader.types.validators.Validator;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@UtilityClass
public class InheritanceLiftUtil {

    public static <ParentT, ValidatableT extends ParentT, FailureT> Validator<ValidatableT, FailureT> liftToChildValidatorType(
            Validator<ParentT, FailureT> parentValidator) {
        return childValidatable -> parentValidator.apply(Either.narrow(childValidatable));
    }

    public static <ParentT, ValidatableT extends ParentT, FailureT> List<Validator<ValidatableT, FailureT>> liftAllToChildValidatorType(
            Collection<Validator<ParentT, FailureT>> parentValidators) {
        return parentValidators.stream()
                .map(InheritanceLiftUtil::<ParentT, ValidatableT, FailureT>liftToChildValidatorType)
                .collect(Collectors.toList());
    }
}

File: src/main/java/org/revcloud/vader/lift/InheritanceLiftUtil.kt
@file:JvmName("InheritanceLiftUtil")

package org.revcloud.vader.lift

import io.vavr.control.Either.narrow
import org.revcloud.vader.types.validators.Validator

/**
 * Assume `ParentT` is parent by inheritance to `ValidatableT`.
 * To validate `ValidatableT`, all it's validations + all its Parent validations has to pass.
 * These utils help to lift it's parent's validations in the context of ValidatableT, so that they can be chained together.
 */
fun <ParentT, ValidatableT : ParentT, FailureT> liftToChildValidatorType(
    parentValidator: Validator<ParentT, FailureT>
): Validator<ValidatableT, FailureT> =
    Validator { childValidatable -> parentValidator.apply(narrow(childValidatable)) }

fun <ParentT, ValidatableT : ParentT, FailureT> liftAllToChildValidatorType(
    parentValidators: Collection<Validator<ParentT, FailureT>>
): List<Validator<ValidatableT, FailureT>> = parentValidators.map { liftToChildValidatorType(it) }
